src/core/MIGRATION.md

```diff
@@ -848,3 +849,43 @@ class MyPlugin {
     this.config$ = initializerContext.config.createIfExists();
   }
 ```
+
+### Mock core services in tests
+Core services already provide mocks to simplify testing and make sure plugins always rely on valid public contracts. 
+```typescript
+// my_plugin/server/plugin.test.ts
+Import { configServiceMock } from 'src/core/server/mocks.ts'
+
+const configService = configServiceMock.create();
+configService.atPath.mockReturnValue(config$);
+…
+const plugin = new MyPlugin({ configService }, …)
+```
+This is not mandatory, but we strongly recommended to export your plugin mocks as well in order for dependent plugins to use them in tests. Your plugin mocks should be exported from the root level of the plugin.
+```typescript
+// my_plugin/server/mocks.ts
+type MyPluginContract = PublicMethodsOf<MyPlugin>;
+
+const createSetupContractMock = () => {
```

Comment:
Using this pattern with this declaration can be overly verbose. In the client mocks, I have been writing them as:

```js
const createSetupContractMock = (): jest.Mocked<PublicMethodsOf<MyPlugin>> = ({
  isValid: jest.fn().mockReturnValue(true),
});
```

No need for temporary variables.

Reply:
in this case type of mock result is ignored because TS doesn't infer type yet
```typescript 
const createSetupContractMock = (): jest.Mocked<PublicMethodsOf<MyPlugin>> = ({
  // wrong type of return value, TS doesn't complain.
  isValid: jest.fn().mockReturnValue({}),
});
```